core/internal/filetransfer/file_transfer_s3.go

```diff
@@ -0,0 +1,262 @@
+	var getObjInput *s3.GetObjectInput = &s3.GetObjectInput{
+		Bucket: aws.String(bucketName),
+		Key:    aws.String(objectName),
+	}
+	var getObjAttrsInput *s3.GetObjectAttributesInput = &s3.GetObjectAttributesInput{
+		Bucket:           aws.String(bucketName),
+		Key:              aws.String(objectName),
+		ObjectAttributes: []types.ObjectAttributes{types.ObjectAttributesEtag},
+	}
+
+	versionId, ok := task.VersionIDString()
+	if ok {
+		getObjAttrsInput.VersionId = &versionId
+		getObjInput.VersionId = &versionId
+	}
+
+	objAttrs, err := ft.client.GetObjectAttributes(ft.ctx, getObjAttrsInput)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the ETag doesn't match what we have stored, try to find the correct version
+	if strings.Trim(*objAttrs.ETag, "\"") != task.Digest {
+		if task.VersionId != nil {
+			return nil, fmt.Errorf(
+				"digest/etag mismatch: etag %s does not match expected digest %s",
+				*objAttrs.ETag,
+				task.Digest,
+			)
+		}
+		getObjInput, err = ft.getCorrectObjectVersion(
+			bucketName,
+			objectName,
+			task.Digest,
+			getObjInput,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return getObjInput, nil
+}
+
+// listObjectsWithPrefix returns a list of all objects in the specified bucket
+// that begin with the given prefix.
+func (ft *S3FileTransfer) listObjectsWithPrefix(
+	bucketName string,
+	prefix string,
+) ([]*s3.GetObjectInput, error) {
+	var objects []*s3.GetObjectInput
+	params := &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucketName),
+		Prefix: aws.String(prefix),
+	}
+	isTruncated := true
+	for isTruncated {
+		output, err := ft.client.ListObjectsV2(ft.ctx, params)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, object := range output.Contents {
+			objects = append(objects, &s3.GetObjectInput{
+				Bucket: aws.String(bucketName),
+				Key:    object.Key,
+			})
+		}
+
+		isTruncated = *output.IsTruncated
+		if isTruncated {
+			params.ContinuationToken = output.NextContinuationToken
+		}
+	}
+	return objects, nil
+}
+
+// getCorrectObjectVersion attempts to find the version of an object with
+// an ETag that matches the specified digest, and returns the input to
+// retrieve that object.
+func (ft *S3FileTransfer) getCorrectObjectVersion(
+	bucketName string,
+	objectName string,
+	digest string,
+	getObjInput *s3.GetObjectInput,
+) (*s3.GetObjectInput, error) {
+	params := &s3.ListObjectVersionsInput{
+		Bucket: aws.String(bucketName),
+		Prefix: aws.String(objectName),
+	}
+	isTruncated := true
+	for isTruncated {
+		versions, err := ft.client.ListObjectVersions(ft.ctx, params)
+		if err != nil {
+			return nil, err
+		}
+		for _, version := range versions.Versions {
+			if strings.Trim(*version.ETag, "\"") == digest {
+				getObjInput.Key = aws.String(*version.Key)
+				getObjInput.VersionId = version.VersionId
+				return getObjInput, nil
+			}
+		}
+		isTruncated = *versions.IsTruncated
+		if isTruncated {
+			params.KeyMarker = versions.NextKeyMarker
+		}
```

Comment:
question/suggestion: (same question as above)
```suggestion
		if output.IsTruncated != nil {
		    isTruncated = *output.IsTruncated
		} else {
		    isTruncated = false
		}
		if isTruncated {
			params.KeyMarker = versions.NextKeyMarker
		}
```